File: 3912.valid-elements-in-an-array/solution.kt
class Solution {
    fun findValidElements(nums: IntArray): List<Int> {
        val rightMax = IntArray(nums.size)
        rightMax[nums.lastIndex] = 0

        for (i in nums.lastIndex - 1 downTo 0) {
            rightMax[i] = maxOf(nums[i + 1], rightMax[i + 1])
        }

        var leftMax = 0
        val res = mutableListOf<Int>()

        for (i in nums.indices) {
            if (nums[i] > leftMax || nums[i] > rightMax[i]) {
                res += nums[i]
            }
            leftMax = maxOf(leftMax, nums[i])
        }

        return res
    }
}

File: 3912.valid-elements-in-an-array/solution.go
func findValidElements(nums []int) []int {
    r := func() []int {
        r := make([]int, len(nums))
        i := len(nums)-1
        r[i] = math.MinInt
        for ; i > 0; i-- {
            r[i-1] = max(nums[i], r[i])
        }
        return r
    }()

    l := math.MinInt
    
    res := make([]int, 0)

    for i, n := range nums {
        if n > l || n > r[i] {
            res = append(res, n)
        }
        l = max(l, n)
    }

    return res
}